valheim-plus-updater

```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+
+# Include defaults and common functions
+. /usr/local/etc/valheim/defaults
+. /usr/local/etc/valheim/common
+
+main() {
+    create_directory_structure
+    check_for_update
+}
+
+create_directory_structure() {
+    mkdir -p "$vp_download_path" "$vp_install_path"
+}
+
+check_for_update() {
+    cd "$vp_download_path"
+    local api_response=$(curl -sL https://api.github.com/repos/nxPublic/ValheimPlus/releases/latest | jq -r ".assets[] | select(.name == \"$vp_zipfile\")")
+    local download_url=$(jq -r  '.browser_download_url' <<< "${api_response}" )
+    local remote_updated_at=$(jq -r  '.updated_at' <<< "${api_response}" )
+    local remote_size=$(jq -r  '.size' <<< "${api_response}")
+
+    if [ -f "$vp_zipfile" -a -f "$vp_zipfile.updated_at" ]; then
+        local_updated_at=$(< "$vp_zipfile.updated_at")
+        local_size=$(stat --format="%s" "$vp_zipfile")
+        if [ "$local_updated_at" = "$remote_updated_at" -a "$local_size" = "$remote_size" ]; then
+            echo "Local ValheimPlus archive is identical to remote archive - no update required"
+        else
+            echo "Local ValheimPlus archive with size $local_size and update date $local_updated_at differs from remote size $remote_size and date $remote_updated_at - updating"
+            prepare_valheim_plus "$download_url" "$remote_updated_at"
+        fi
+    else
+        echo "Fresh ValheimPlus install"
+        prepare_valheim_plus "$download_url" "$remote_updated_at"
+    fi
+    if [ -f "$vp_mergefile" ]; then
+        echo "Valheim dedicated server or ValheimPlus got updated - merging installation files"
+        rm -f "$vp_mergefile"
+        merge_valheim_plus
+    fi
+}
+
+prepare_valheim_plus() {
+    local download_url="$1"
+    local updated_at="$2"
+    download_valheim_plus "$download_url" $updated_at \
+        && extract_valheim_plus \
+        && touch "$vp_mergefile"
+}
+
+merge_valheim_plus() {
+    rm -rf "$vp_install_path.tmp" "$vp_install_path.old"
+    mkdir -p "$vp_install_path.tmp"
+    rsync -a --itemize-changes --exclude server_exit.drp --exclude steamapps "$valheim_download_path/" "$vp_install_path.tmp"
+    rsync -a --itemize-changes "$vp_download_path/extracted/" "$vp_install_path.tmp"
+    if [ ! -d "/config/valheimplus" ]; then
+        cp -a "$vp_install_path.tmp/BepInEx/config" "/config/valheimplus"
+    fi
+    rm -rf "$vp_install_path.tmp/BepInEx/config"
+    chmod +x "$vp_install_path.tmp/start_game_bepinex.sh"
+    ln -s /config/valheimplus "$vp_install_path.tmp/BepInEx/config"
```

Comment:
Should we `chmod` this directory so that it can be writeable by the user? Otherwise, only `root` can write to the `valheim_plus.cfg` file

Reply:
I was contemplating introducing new env variables for those config file permissions. But then figured I'd also want to `chown` to another UID potentially. And then I figured if it was an issue we'd already have it for the existing `/config` files which also get created as UID `0` if the user doesn't supply them. Like, if the user volume mounts existing config it won't be touched. So I wasn't sure what to do here. But maybe adding the ability to `chmod` the files is a good start.

Reply:
Okay, introduced `VALHEIM_PLUS_CONFIG_DIRECTORY_PERMISSIONS` and `VALHEIM_PLUS_CONFIG_FILE_PERMISSIONS`